### PR TITLE
ROS-247 and ROS-314: Create primitive definitions located in Queue.cc

### PR DESCRIPTION
**Pull request by linsheng9731**
This PR related to these Jira tickets.
https://rchain.atlassian.net/browse/ROS-247
https://rchain.atlassian.net/browse/ROS-314

### CODE REVIEW

**Comment by dckc** [comment]
This looks good at a glance, @linsheng9731. I expect someone from the core team will deal with it in a few days to a few weeks. I haven't looked closely enough to approve it.

I see that it does include tests, so I'm surprised the coverage tool doesn't think it will affect coverage at all. I haven't checked the travis logs to see whether the tests are getting run.

**Comment by linsheng9731** [comment]
@dckc Thank you. It seems that test coverage not enabled in roscala project, but i have already run the tests and guaranteed at least 88% coverage.

**Comment by rabbitonweb** [comment]
I will review it from the code perspective, someone should have a look from domain perspective (@tymm ?)

**Comment by tymm** [comment]
@rabbitonweb @linsheng9731 Yes, I'll also look at it.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, RblBool] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      Right(RblBool(q.empty()))
+    }
+  }
+
+
+  object QueueEnQueue extends Prim {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: `QueueEnqueue`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  object QueueEmpty extends Prim {
+    override val name: String = "queue-empty?"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, RblBool] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      Right(RblBool(q.empty()))
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Is there a reason you didn't name the function `isEmpty`?

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    override val name: String = "queue-enqueue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Queue] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      val ob = ctxt.argvec.elem(1)
+      q.enQueue(ob)
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Please rename to `enqueue`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+  object QueueEnQueue extends Prim {
+    override val name: String = "queue-enqueue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Queue] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Please check if the first element is actually a `Queue` and if not return `Left` with the appropriate error.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    }
+  }
+
+
+  object QueueEnQueue extends Prim {
+    override val name: String = "queue-enqueue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Remove the `checkTypeMismatch[Ob]`. Doesn't really do anything in this case.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+      Right(q)
+    }
+  }
+
+  object QueueDeQueue extends Prim {
+    override val name: String = "queue-dequeue"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Ob]
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Please remove.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+      Right(q.head)
+    }
+  }
+
+  object QueuePatDeQueue extends Prim {
+    override val name: String = "queue-pat-dequeue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Please remove.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+      }
+    }
+  }
+
+  object QueuePatRead extends Prim {
+    override val name: String = "queue-pat-read"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Please remove.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+      }
+    }
+  }
+
+  object QueueReadNth extends Prim {
+    override val name: String = "queue-read-nth"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Please remove.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+      Right(q.elems.elem(n))
+    }
+  }
+
+  object QueueDequeueNth extends Prim {
+    override val name: String = "queue-dequeue-nth"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Please remove.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Queue] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      val ob = ctxt.argvec.elem(1)
+      q.enQueue(ob)
+      Right(q)
+    }
+  }
+
+  object QueueDeQueue extends Prim {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: `queueDequeue`

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    override val name: String = "queue-dequeue"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Ob]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Please check if the first element is actually a `Queue` and if not return `Left` with the appropriate error.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+      if(q.empty()) {
+        Left(QueueEmptyError("queue is empty"))
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: The message inside of the `QueueEmptyError` is not really providing any additional value. Can you change the `QueueEmptyError` to a `case object`?

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+  object QueueDeQueue extends Prim {
+    override val name: String = "queue-dequeue"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Ob]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      if(q.empty()) {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: `if else` please.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+  object QueueRead extends Prim {
+    override val name: String = "queue-read"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      if(q.empty()) {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: `if else` please.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+
+  object QueueRead extends Prim {
+    override val name: String = "queue-read"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Please check if the first element is actually a `Queue` and if not return `Left` with the appropriate error.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+      if(q.empty()) {
+        Left(QueueEmptyError("queue is empty"))
+      }
+      Right(q.head)
+    }
+  }
+
+  object QueuePatDeQueue extends Prim {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: `queuePDequeue`

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+    override val name: String = "queue-pat-dequeue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Please check if the first element is actually a `Queue` and if not return `Left` with the appropriate error.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+      val pat = ctxt.argvec.elem(1).asInstanceOf[Tuple]
+      if(q.empty()) {
+        Left(QueueEmptyError("queue is empty"))
+      } else if(pat == Tuple.NIL) {
+        Right(q.deQueue())
+      } else {
+        Right(q.patternDequeue(pat).getOrElse(ABSENT))
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Return `Left` with appropriate error if `q.patternDequeue(pat)` is `None`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+    override val name: String = "queue-pat-dequeue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      val pat = ctxt.argvec.elem(1).asInstanceOf[Tuple]
+      if(q.empty()) {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Remove braces in `if` statement.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  object QueuePatRead extends Prim {
+    override val name: String = "queue-pat-read"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: See comments from previous primitive.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  object QueueReadNth extends Prim {
+    override val name: String = "queue-read-nth"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Ob]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: See previous comments.

File: roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+  object QueueReset extends Prim {
+    override val name: String = "queue-reset"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Queue] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      q.reset()
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/PrimQueue.scala]: Seems like `reset` has side-effects. Please make it side-effect free.

File: roscala/src/main/scala/coop/rchain/rosette/Queue.scala [hunk 24]
```diff
@@ -0,0 +1,111 @@
+package coop.rchain.rosette
+
+import java.util.NoSuchElementException
+
+case class Queue(var elems: Tuple) extends Ob {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/Queue.scala]: Please don't use mutation in `Queue`.

File: roscala/src/main/scala/coop/rchain/rosette/Queue.scala [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+
+  def empty(): Boolean =
+    (nElem.asInstanceOf[Fixnum] == Fixnum(0)).value
+
+  def enQueue(ele: Ob): Unit =
+    elems = Tuple(elems.elem :+ ele)
+
+  def deQueue(): Ob =
+    if (empty()) {
+      throw new NoSuchElementException("queue empty")
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/Queue.scala]: We don't want to throw exceptions.

**Comment by dckc** [comment]
@linsheng9731 be sure to see the [Coding Standards][1], linked from CONTRIBUTING.md. For example:

> Scala provides an exception facility, but do not use it for commonplace errors ...

[1]: https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards

**Comment by linsheng9731** [comment]
@tymm Thank you very much! I have followed your suggestions and made some changes. Please review it again!

**Comment by linsheng9731** [comment]
After i rebased dev branch,  there was something wrong with this PR. So i made a new pull-request here.  https://github.com/rchain/rchain/pull/296. Please close this one.

**Pull request closed without merging**